### DISCUSSION TITLE
How do proxy settings apply in VSCode extension?

### DISCUSSION THREAD

**Discussion by alekdavis** [AnWeber/vscode-httpyac#346] [General]
I spent a couple of days troubleshooting issues and helping our devs, and in my observation, majority of people who could not make it work and gave up, did so because they could not make the proxy configuration work.

In CLI, it looks like the proxy is recognized in the `httpyac.config.js` (`proxy` element) and `request_proxy` (environment) variable (this one overrides the other). I got a report form another dev that on a Mac, the `httpyac.config.js` does not work, but I cannot validate because I do not have a Mac. But the CLI case is more or less clear. I thought that CLI also recognizes `http_proxy`, but in my tests, it did not (not a big deal, maybe it's good that it doesn't, since fewer options make it simpler).

The VSCode case is more confusing. VSCode has its own proxy settings, and if not defined, it recognizes `http_proxy`. Then, I assume, you can also use the `httpyac.config.js` (`proxy` element) and `request_proxy` (environment) variable, right? I wanted to test it, but I cleared all settings (so, no `http_proxy`, no VSCode proxy definition, no proxy setting in the `httpyac.config.js` file, and no `request_proxy` (environment) variable), but I can still make calls that require a proxy and they work just fine. I am wondering if the extension recognizes the default system proxy (mine points to a corporate script that has about 100 lines of code to determine proxy settings based on the endpoint URL), so I doubt that VSCode is that smart, but I cannot explain why it would work when no other proxy definitions are set.

Another thing we found was that VSCode also recognizes the `no_proxy` environment variable which messed us badly because on some of our dev systems, it has a setting that holds the company domain, but there are some APIs under the company domain that are hosted externally, so they need to go through a proxy. That's mostly our issue (I'm not sure why our IT did this), but it took us a few hours to figure out that the `no_proxy` setting was responsible (I did not see anywhere that VSCode can use it). 

It would be really helpful, if a clear description of how the proxy settings work.

**Comment by AnWeber** [comment]
Hello, 
I will first explain what httpyac offers in terms of proxy settings.I have recently been trying to detect an HTTP_PROXY or HTTPS_PROXY environment variable and would then apply it. Under VSCode, the extension also accesses the `http.proxy` setting and applies it if available. To activate the proxy explicitly, it must be set explicitly on the request (via script, request_proxy variable or `# @proxy ...`). If some of these options are enabled I will create a [Http Proxy Agent](https://www.npmjs.com/package/http-proxy-agent) or [Socks Proxy Agent](https://www.npmjs.com/package/socks-proxy-agent) determined by the proxy protocol.

VSCode uses the same Proxy settings as Chromium like described in [docs](https://code.visualstudio.com/docs/setup/network#_proxy-server-support). Then there is this [magical setting 'http.proxySupport'](https://code.visualstudio.com/docs/getstarted/settings). This overrides all proxy settings for VSCode extensions and activates a separate proxy agent (which cannot use http2). As long as the setting is not set to off, all settings in httpyac are ignored, or so it seems to me. 

To understand problems with proxy in vscode you can activate extended logging via CLI ('code --log-net-log=netlog.json' ). Maybe this will help you.
https://gist.github.com/PSJoshi/0831f635f9f5a7b9c602b80fbbf3125a#tracking-network-issues-while-running-vs-code-editor

**Reply by alekdavis** [reply]
Ah, I did not realize VS Code understands the browser proxy settings. That would explain why mine works without anything configured in VS Code. Also, didn't know about `request_proxy` and `# @ proxy` these will be handy for troubleshooting. Thanks a lot for the explanation. It makes more sense now.
